File: #109_Convert_Sorted_List_to_Binary_Search_Tree/Solution.java
class ListNode {
    int val;
    ListNode next;
    ListNode(int x) { val = x; }
 }

class TreeNode {
    int val;
    TreeNode left;
    TreeNode right;
    TreeNode(int x) { val = x; }
}

public class Solution {
	ListNode node;
    public TreeNode sortedListToBST(ListNode head) {
        if (head == null) return null;
        node = head;
        int len = countList(head);
        return build(0, len - 1);
    }

    private TreeNode build(int lo, int hi) {
    	// System.out.println("lo: " + lo + "hi: " + hi);
    	if (lo > hi) return null;
    	int mid = (lo + hi) / 2;
    	TreeNode left = build(lo, mid - 1);
    	TreeNode center = new TreeNode(node.val);
    	node = node.next;
    	center.left = left;
    	center.right = build(mid + 1, hi);
    	return center;
    }

    private int countList(ListNode head) {
    	if (head == null) return 0;
    	int count = 0;
    	ListNode nd = head;
    	while (nd != null) {
    		count++;
    		nd = nd.next;
    	}
    	return count;
    }

    public static void main(String[] args) {
    	ListNode head = new ListNode(0);
    	Solution sol = new Solution();
    	sol.sortedListToBST(head);
    }
}

File: #109_Convert_Sorted_List_to_Binary_Search_Tree/Solution.py
'''
In order traversal solution
'''
class ListNode:
	def __init__(self, x):
		self.val = x
		self.next = None

class TreeNode:
	def __init__(self, x):
		self.val = x
		self.left = None
		self.right = None

class Solution:

	def __init__(self):
		self.cur = None

	# @param {ListNode} head
	# @return {TreeNode}
	def sortedListToBST(self, head):
		self.cur = head
		return self.build(self.countList(head))

	def build(self, n):
		if n < 1:
			return None
		root = TreeNode(0)
		root.left = self.build(n / 2)
		root.val = self.cur.val
		self.cur = self.cur.next
		root.right = self.build(n - n / 2 - 1)
		return root

	def countList(self, head):
		ans = 0
		cur = head
		while cur:
			ans += 1
			cur = cur.next
		return ans